mongo-store.js

```diff
@@ -180,79 +181,189 @@ module.exports = function (opts) {
+          return
+        }
 
-      var update = !!ent.id
+        const is_update = null != msg.ent.id
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+        if (is_update) {
+          return updateExisting(msg, coll, done)
+        }
+
+        return createAndSave(msg, coll, done)
+      })
+
+
+      function createAndSave(msg, coll, done) {
+        if (isUpsertRequested(msg)) {
+          return doUpsert(msg, coll, done)
+        }
+
+        return createNew(msg, coll, done)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+        function isUpsertRequested(msg) {
+          if (null == msg.q) {
+            return false
           }
 
-          entp._id = makeid(id)
+          const public_entdata = msg.ent.data$(false)
+          const upsert_on = msg.q.upsert$
 
-          coll.insertOne(entp, function (err, inserts) {
-            if (!error(args, err, cb)) {
-              var entu = inserts.ops[0]
-              var fent = null
-              if (entu) {
-                entu.id = idstr(entu._id)
-                delete entu._id
-                //fent = ent.make$(_.cloneDeep(entu))
-                fent = ent.make$(seneca.util.deep(entu))
+          return Array.isArray(upsert_on) &&
+            upsert_on.length > 0 &&
+            upsert_on.every(p => p in public_entdata)
+        }
+
+        function doUpsert(msg, coll, done) {
+          const upsert_on = msg.q.upsert$
+          const public_entdata = msg.ent.data$(false)
+
+          const filter_by = upsert_on
+            .reduce((acc, field) => {
+              acc[field] = msg.ent[field]
+              return acc
+            }, {})
+
+          const replacement = (() => {
+            const o = Dot.flatten(Object.assign({}, public_entdata))
+
+
+            const NO_SPECIFIC_ID_REQUESTED = {}
+
+            const id = tryMakeIdIfRequested(msg, {
+              when_no_specific_id_requested: NO_SPECIFIC_ID_REQUESTED
+            })
+
+
+            if (id !== NO_SPECIFIC_ID_REQUESTED) {
+              o.$setOnInsert = { _id: id }
+            }
+
+            return o
+          })()
+
+          return coll.findOneAndUpdate(
+            filter_by,
+            replacement,
+            { upsert: true, returnNewDocument: true },
+
+            function (err, entu) {
+              if (error(msg, err, done)) {
+                return
               }
-              seneca.log.debug('save/insert', ent, desc)
-              cb(null, fent)
+
+              return done(null, entu)
+            }
+          )
+        }
+
+
+        function createNew(msg, coll, done) {
+          const new_doc = (function () {
+            const public_entdata = msg.ent.data$(false)
+
+
+            const NO_SPECIFIC_ID_REQUESTED = {}
```

Comment:
Abstract id handing into a utility function called `ent = ensure_id(ent)`. All it needs to do is:
```
          if (undefined !== ent.id$) {
            id = ent.id$
          } else if (opts.generate_id) {
            id = opts.generate_id(ent)
          }

          entp._id = makeid(id)
```

Also be careful of needlessly copying data via Object.assign

Reply:
Okay, I will extract the code, you have specified in the snippet, into a separate helper called `ensure_id`.